### PR TITLE
Send PEM files to server when provisioning a log. 

### PR DESCRIPTION
**Pull request by gdbelvin**
The admin client currently only supports provisioning logs and maps with keys that are locally on disk.  Since the files being supplied by the admin client may not already be on the server's local disk, this presents a bit of a tricky puzzle. 

This PR sends the contents of the private key and the server stores the key directly in the admin db. 

This PR also removes the requirement for PEM files to have passwords since the server does not have this requirement.

### CODE REVIEW

File: cmd/createtree/main.go [hunk 1]
```diff
@@ -141,16 +142,28 @@ func newPK(opts *createOpts) (*any.Any, error) {
 	switch opts.privateKeyType {
 	case "PEMKeyFile":
 		if opts.pemKeyPath == "" {
-			return nil, errors.New("empty PEM path")
```
Comment (codingllama) [inline cmd/createtree/main.go]: Please keep the checks - they are useful for tool users (and slightly more readable than the resulting server-side error message they would get).

File: cmd/createtree/main.go [hunk 2]
```diff
@@ -141,16 +142,28 @@ func newPK(opts *createOpts) (*any.Any, error) {
...
 			Path:     opts.pemKeyPath,
 			Password: opts.pemKeyPass,
 		}
 		return ptypes.MarshalAny(pemKey)
+	case "PrivateKey":
+		if opts.pemKeyPath == "" {
+			return nil, errors.New("empty pem_key_path")
+		}
+		pemSigner, err := keys.NewFromPrivatePEMFile(
+			opts.pemKeyPath, opts.pemKeyPass)
```
Comment (codingllama) [inline cmd/createtree/main.go]: It seems conceptually strange to me to force a file to exist in this case. We could allow for a literal key to be passed via a flag. That possibly means the key gets logged on shell history, unless the user is careful, so there's that as well. I'm not too sure TBH.

WDYT?

File: crypto/keys/pem_signer_factory.go [hunk 3]
```diff
@@ -72,10 +71,11 @@ func (f PEMSignerFactory) Generate(ctx context.Context, tree *trillian.Tree, spe
 		return nil, fmt.Errorf("Specification produces %T, but tree.SignatureAlgorithm = %v", key, tree.GetSignatureAlgorithm())
 	}
 
-	return marshalPrivateKey(key)
+	return MarshalPrivateKey(key)
 }
 
-func marshalPrivateKey(key crypto.Signer) (*any.Any, error) {
+// MarshalPrivateKey returns an any.Any with a keyspb.PrivateKey inside.
```
Comment (codingllama) [inline crypto/keys/pem_signer_factory.go]: I think this would be better placed at `private.go` now, considering that we've just made it public. I would also prefer for it to have its own set of tests.

File: cmd/createtree/main.go [hunk 4]
```diff
@@ -141,16 +142,28 @@ func newPK(opts *createOpts) (*any.Any, error) {
...
 		pemKey := &keyspb.PEMKeyFile{
 			Path:     opts.pemKeyPath,
 			Password: opts.pemKeyPass,
 		}
 		return ptypes.MarshalAny(pemKey)
+	case "PrivateKey":
+		if opts.pemKeyPath == "" {
+			return nil, errors.New("empty pem_key_path")
+		}
+		pemSigner, err := keys.NewFromPrivatePEMFile(
```
Comment (codingllama) [inline cmd/createtree/main.go]: Please add a `pemKeyPass == ""` check here too, as above. (Reasoning in another comment.)

File: cmd/createtree/main_test.go [hunk 5]
```diff
@@ -87,6 +102,9 @@ func TestRun(t *testing.T) {
 	emptyPEMPass := *validOpts
 	emptyPEMPass.pemKeyPass = ""
 
+	pemkey := *validOpts
```
Comment (codingllama) [inline cmd/createtree/main_test.go]: nit: pemKey

File: crypto/keys/private.go [hunk 6]
```diff
@@ -56,9 +56,6 @@ type SignerFactory interface {
 // NewFromPrivatePEMFile reads a PEM-encoded private key from a file.
 // The key must be protected by a password.
 func NewFromPrivatePEMFile(keyFile, keyPassword string) (crypto.Signer, error) {
-	if keyPassword == "" {
```
Comment (codingllama) [inline crypto/keys/private.go]: Please don't remove this check, we've added it on purpose. The intention is to foster good practices along with Trillian usage. People really shouldn't be storing unencrypted PEM files in their servers.

There's some context at #544. If you feel strongly about this, please move the discussion to a separate PR.

File: cmd/createtree/main.go [hunk 7]
```diff
@@ -141,16 +142,28 @@ func newPK(opts *createOpts) (*any.Any, error) {
...
-		if opts.pemKeyPass == "" {
-			return nil, fmt.Errorf("empty password for PEM key file %q", opts.pemKeyPath)
+			return nil, errors.New("empty pem_key_path")
 		}
 		pemKey := &keyspb.PEMKeyFile{
 			Path:     opts.pemKeyPath,
 			Password: opts.pemKeyPass,
 		}
 		return ptypes.MarshalAny(pemKey)
+	case "PrivateKey":
```
Comment (codingllama) [inline cmd/createtree/main.go]: Would mind making this the default (line 61)? @RJPercival and I discussed this a bit, and we both seem to prefer it slightly over PEMKeyFile. (I'm guessing you do too.)

File: cmd/createtree/main.go (hunk 1 above)
Comment (gdbelvin) [inline cmd/createtree/main.go]: adding checks back in

File: cmd/createtree/main.go (hunk 7 above)
Comment (gdbelvin) [inline cmd/createtree/main.go]: done

File: cmd/createtree/main.go (hunk 4 above)
Comment (gdbelvin) [inline cmd/createtree/main.go]: done

File: cmd/createtree/main.go (hunk 2 above)
Comment (gdbelvin) [inline cmd/createtree/main.go]: I think it's probably better to focus on generating the key server side.

File: cmd/createtree/main_test.go (hunk 5 above)
Comment (gdbelvin) [inline cmd/createtree/main_test.go]: done

File: crypto/keys/pem_signer_factory.go (hunk 3 above)
Comment (gdbelvin) [inline crypto/keys/pem_signer_factory.go]: moved

File: crypto/keys/private.go (hunk 6 above)
Comment (gdbelvin) [inline crypto/keys/private.go]: added back

**Comment by gdbelvin** [comment]
@codingllama or @RJPercival Would you mind reviewing?

File: crypto/keys/keys_test.go [hunk 8]
```diff
@@ -324,3 +324,25 @@ func TestGenerateKey(t *testing.T) {
 		}
 	}
 }
+
+func TestMarshalPrivateKey(t *testing.T) {
+	for _, tc := range []struct {
+		pem     string
+		typeURL string
+	}{
+		{ecdsaPrivateKey, "type.googleapis.com/keyspb.PrivateKey"},
```
Comment (codingllama) [inline crypto/keys/keys_test.go]: nit: Please use field keys, it makes future changes less painful.

`{pem: ecdsaPrivateKey, typeURL: "..."},`

File: crypto/keys/keys_test.go [hunk 9]
```diff
@@ -324,3 +324,25 @@ func TestGenerateKey(t *testing.T) {
...
+		{rsaPrivateKey, "type.googleapis.com/keyspb.PrivateKey"},
+	} {
+		k, err := NewFromPrivatePEM(tc.pem, "")
+		if err != nil {
+			t.Errorf("NewFromPrivatePEM(%v): %v", tc.pem, err)
```
Comment (codingllama) [inline crypto/keys/keys_test.go]: Add `continue` after t.Errorf, otherwise it'll probably panic on failed tests. Ditto for others.

File: crypto/keys/keys_test.go [hunk 10]
```diff
@@ -324,3 +324,25 @@ func TestGenerateKey(t *testing.T) {
...
+		}
+		a, err := MarshalPrivateKey(k)
+		if err != nil {
+			t.Errorf("MarshalPrivateKey(%v): %v", k, err)
+		}
+		if got, want := a.TypeUrl, tc.typeURL; got != want {
```
Comment (codingllama) [inline crypto/keys/keys_test.go]: We should check if the DER on the PrivateKey proto is correct, otherwise the test isn't adding much.

File: crypto/keys/private_keys.go [hunk 11]
```diff
@@ -159,3 +160,23 @@ func curveFromParams(params *keyspb.Specification_ECDSA) elliptic.Curve {
...
+
+	switch key := key.(type) {
+	case *ecdsa.PrivateKey:
+		privateKeyDER, err = x509.MarshalECPrivateKey(key)
+	case *rsa.PrivateKey:
+		privateKeyDER = x509.MarshalPKCS1PrivateKey(key)
+	default:
+		return nil, fmt.Errorf("unsupported key type: %T", key)
+	}
+	if err != nil {
```
Comment (codingllama) [inline crypto/keys/private_keys.go]: nit: Move `err`'s declaration and the if inside the ecdsa case? Neither need to be at this scope.

File: cmd/createtree/main_test.go [hunk 12]
```diff
@@ -134,6 +149,32 @@ func TestRun(t *testing.T) {
 			desc:    "emptyPEMPass",
 			opts:    &emptyPEMPass,
 			wantErr: true,
+			wantTree: &trillian.Tree{
```
Comment (gdbelvin) [inline cmd/createtree/main_test.go]: done

File: crypto/keys/keys_test.go (hunk 8 above)
Comment (gdbelvin) [inline crypto/keys/keys_test.go]: done

File: crypto/keys/keys_test.go (hunk 9 above)
Comment (gdbelvin) [inline crypto/keys/keys_test.go]: done

File: crypto/keys/keys_test.go (hunk 10 above)
Comment (gdbelvin) [inline crypto/keys/keys_test.go]: I added a check to make sure the type contained in the DER was correct. I hope that's close enough?

File: crypto/keys/private_keys.go (hunk 11 above)
Comment (gdbelvin) [inline crypto/keys/private_keys.go]: done

File: crypto/keys/keys_test.go (hunk 10 above)
Comment (gdbelvin) [inline crypto/keys/keys_test.go]: Actually, a rebase removed the need for this method.

**Comment by gdbelvin** [comment]
A rebase simplified this PR. PTAL

File: cmd/createtree/main.go (hunk 2 above)
Comment (RJPercival) [inline cmd/createtree/main.go]: I think it's best to limit it to files since, as you point out, there are risks to allowing a private key to appear on the command-line. Besides, if someone doesn't want it to touch disk, they can do something like:
```shell
openssl ecparam -name secp256k1 -genkey -noout | createtree --pem_key_path /dev/stdin --pem_key_password "foo"
```

File: cmd/createtree/main_test.go [hunk 13]
```diff
@@ -46,7 +58,7 @@ func TestRun(t *testing.T) {
 		HashStrategy:       trillian.HashStrategy_RFC_6962,
 		HashAlgorithm:      sigpb.DigitallySigned_SHA256,
 		SignatureAlgorithm: sigpb.DigitallySigned_RSA,
-		PrivateKey:         anyKey,
+		PrivateKey:         marshalAny(&keyspb.PrivateKey{Der: pemDer}),
```
Comment (RJPercival) [inline cmd/createtree/main_test.go]: nit: It might be nicer to still do the marshaling at the start of the test, so you can report an error using `t.Fatalf()`.

File: cmd/createtree/main_test.go [hunk 14]
```diff
@@ -136,6 +151,22 @@ func TestRun(t *testing.T) {
 			wantErr: true,
 		},
 		{
+			desc:    "PEMKeyFile",
+			opts:    &pemKeyOpts,
+			wantErr: false,
+			wantTree: &trillian.Tree{
```
Comment (RJPercival) [inline cmd/createtree/main_test.go]: nit: Making this tree at the start of the test might be nicer, in the same way that nonDefaultTree was created. Then you could get rid of `marshalAny()` as well.

File: cmd/createtree/main.go (hunk 2 above)
Comment (RJPercival) [inline cmd/createtree/main.go]: Ah actually, that wouldn't work, because we require that the private key has a password. There's bound to be an openssl command to put in the middle that will (somewhat pointlessly) encrypt the key though.

File: cmd/createtree/main_test.go (hunk 13 above)
Comment (gdbelvin) [inline cmd/createtree/main_test.go]: `marshalAny` is used twice.  I could remove it and do all of the marshaling before the test to take advantage of `t.Fatalf`, but my thought was that the values we are marshaling are test configuration values which won't change as a result of code, and that it was nice to be able to use the output of `marshalAny` directly in building the table driven tests.

File: cmd/createtree/main_test.go (hunk 14 above)
Comment (gdbelvin) [inline cmd/createtree/main_test.go]: Good idea. done.
